### PR TITLE
Ben Donaldson

### PR DESCRIPTION
**Pull request by bendonaldson**

### CODE REVIEW

File: brevets/templates/calc.html [hunk 1]
```diff
@@ -1,176 +1,173 @@
...
-            $(document).ready(function(){
-             // Do the following when the page is finished loading
+            $.getJSON(TIME_CALC_URL, { km: km, distance: distance, begin_date: begin_date }).done(
+               function (data) {
+                  var times = data.result;
+                  console.log("Got a response: ");
+                  console.log("GOT HERE IN THE TESTING")
+                  console.log("Response.open = " + times.open);
+                  // This part will automatically fill the open and close slots,
+                  // so don't worry about those.
```
Comment (vlanier2) [inline brevets/templates/calc.html:129]: Front end looks good and everything is working on the webpage when I tested it.

File: brevets/acp_times.py [hunk 2]
```diff
@@ -41,4 +79,43 @@ def close_time(control_dist_km, brevet_dist_km, brevet_start_time):
...
+        if dist <= distance:
+            time += dist / min_speeds[speed]
+            distance -= dist
+        else:
+            time += (control_dist_km - speed[0]) / min_speeds[speed]
+            distance -= control_dist_km - speed[0]
+            break
+
+    hour, mins = convert_time(time)
+    return brevet_start_time.shift(hours=+hour, minutes=+mins)
```
Comment (vlanier2) [inline brevets/acp_times.py:112]: I like how the edge cases are organized. Same comment as the other loop. The convert_time helper is really putting in work here.

File: brevets/flask_brevets.py [hunk 3]
```diff
@@ -51,14 +51,12 @@ def _calc_times():
...
-    # FIXME!
-    # Right now, only the current time is passed as the start time
-    # and control distance is fixed to 200
-    # You should get these from the webpage!
-    open_time = acp_times.open_time(km, 200, arrow.now().isoformat).format('YYYY-MM-DDTHH:mm')
-    close_time = acp_times.close_time(km, 200, arrow.now().isoformat).format('YYYY-MM-DDTHH:mm')
+    open_time = acp_times.open_time(km, distance, arrow.get(begin_date)).format('YYYY-MM-DDTHH:mm')
+    close_time = acp_times.close_time(km, distance, arrow.get(begin_date)).format('YYYY-MM-DDTHH:mm')
     result = {"open": open_time, "close": close_time}
     return flask.jsonify(result=result)
```
Comment (vlanier2) [inline brevets/flask_brevets.py:61]: Everything looks good on the flask side

File: brevets/tests/test_acp_times.py [hunk 4]
```diff
@@ -6,6 +6,69 @@
...
+
+def test_close_time_1000km():
+    assert arrow.get("2022-01-01T01:00") == acp_times.close_time(0, 1000, arrow.get("2022-01-01T00:00"))
+    assert arrow.get("2022-01-01T10:00") == acp_times.close_time(150, 1000, arrow.get("2022-01-01T00:00"))
+    assert arrow.get("2022-01-01T20:00") == acp_times.close_time(300, 1000, arrow.get("2022-01-01T00:00"))
+    assert arrow.get("2022-01-02T06:00") == acp_times.close_time(450, 1000, arrow.get("2022-01-01T00:00"))
+    assert arrow.get("2022-01-02T16:00") == acp_times.close_time(600, 1000, arrow.get("2022-01-01T00:00"))
+    assert arrow.get("2022-01-03T05:08") == acp_times.close_time(750, 1000, arrow.get("2022-01-01T00:00"))
+    assert arrow.get("2022-01-03T18:15") == acp_times.close_time(900, 1000, arrow.get("2022-01-01T00:00"))
+    assert arrow.get("2022-01-04T03:00") == acp_times.close_time(1000, 1000, arrow.get("2022-01-01T00:00"))
```
Comment (vlanier2) [inline brevets/tests/test_acp_times.py:74]: The test cases are reasonable and everything passed. It makes sense to separate the open tests from the close tests since those are separate functions. Some of the edge cases are not tested here I think. I don't see a test for a control time that is larger than the total brevet distance. You could push the code a bit harder with bunch of small or large values to try and find an inconsistency with the ACP calculator.

**Review by vlanier2 — COMMENTED** [review]
Hi Ben, 

It was great to review your project this week. I like some of your design decisions. Especially the use of a time conversion helper function and the way you split up the speed tables. It really cleans up the code and makes it shorter. Everything was understandable but I would recommend a bit more descriptive variable names for ease of reading later on. 

The front end and flask are clear and as described in class. 

All the tests passed and I ran through a few brevets on the page, everything matched the ACP calculator. I think the testing could be more aggressive to catch possible errors close to the boundaries. Like some controls at 1, 2, 1199, 1200 etc.

I hope you had luck with future projects!
